### PR TITLE
[#4832] feat(core): Add basic framework to support multiple JDBC backend.

### PR DESCRIPTION
**Pull request by yuqi1129**
### What changes were proposed in this pull request?

1. Add a framework to support multiple JDBC backend 
2. Modify all the mapper classes related to metalake, catalog, schema, filesetset, table, topic and use new framework, others will use another PR to avoid a large PR.

### Why are the changes needed?

To support more JDBC storage backend.

Fix: #4832 

### Does this PR introduce _any_ user-facing change?

N/A

### How was this patch tested?

Existing tests.

### CODE REVIEW

File: core/src/main/java/org/apache/gravitino/storage/relational/mapper/CatalogMetaSQLProvider.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+    JDBCBackendType jdbcBackendType = JDBCBackendType.fromString(databaseId);
+    return METALAKE_META_SQL_PROVIDER_MAP.get(jdbcBackendType);
+  }
+
+  static class CatalogMetaMySQLProvider extends CatalogMetaBaseProvider {}
+
+  static class CatalogMetaH2Provider extends CatalogMetaBaseProvider {}
+
+  static class CatalogMetaPostgreSQLProvider extends CatalogMetaBaseProvider {}
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/storage/relational/mapper/CatalogMetaSQLProvider.java:53]: Remove the PG part in this PR, they can be done in the separate PR.

File: core/src/main/java/org/apache/gravitino/storage/relational/mapper/CatalogMetaSQLProvider.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+package org.apache.gravitino.storage.relational.mapper;
+
+import com.google.common.collect.ImmutableMap;
+import java.util.Map;
+import org.apache.gravitino.storage.relational.JDBCBackend.JDBCBackendType;
+import org.apache.gravitino.storage.relational.po.CatalogPO;
+import org.apache.gravitino.storage.relational.session.SqlSessionFactoryHelper;
+import org.apache.ibatis.annotations.Param;
+
+public class CatalogMetaSQLProvider {
```
[the remark this answers is not in the document]
Comment (yuqi1129) [reply core/src/main/java/org/apache/gravitino/storage/relational/mapper/CatalogMetaSQLProvider.java:29]: I have rename `CatalogMetaSQLProvider` to `CatalogMetaProviderFactory` and try to make all methods in `CatalogMetaProviderFactory` are static.  Currently `CatalogMetaSQLProvider` contains some object-level method and I'm afraid that it's not a very good idea to merge it with `CatalogMetaBaseProvider`.

File: core/src/main/java/org/apache/gravitino/storage/relational/mapper/CatalogMetaSQLProvider.java (hunk 1 above)
Comment (yuqi1129) [reply core/src/main/java/org/apache/gravitino/storage/relational/mapper/CatalogMetaSQLProvider.java:53]: Done

Comment (yuqi1129) [reply core/src/main/java/org/apache/gravitino/storage/relational/mapper/CatalogMetaSQLProvider.java]: done

File: core/src/main/java/org/apache/gravitino/storage/relational/mapper/CatalogMetaBaseProvider.java [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        + " catalog_comment as catalogComment, properties, audit_info as auditInfo,"
+        + " current_version as currentVersion, last_version as lastVersion,"
+        + " deleted_at as deletedAt"
+        + " FROM "
+        + TABLE_NAME
+        + " WHERE metalake_id = #{metalakeId} AND deleted_at = 0";
+  }
+
+  public String selectCatalogIdByMetalakeIdAndName(
+      @Param("metalakeId") Long metalakeId, @Param("catalogName") String named) {
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/storage/relational/mapper/CatalogMetaBaseProvider.java:40]: what is "named" here?

File: core/src/main/java/org/apache/gravitino/storage/relational/mapper/CatalogMetaBaseProvider.java [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        + " audit_info = #{newCatalogMeta.auditInfo},"
+        + " current_version = #{newCatalogMeta.currentVersion},"
+        + " last_version = #{newCatalogMeta.lastVersion},"
+        + " deleted_at = #{newCatalogMeta.deletedAt}"
+        + " WHERE catalog_id = #{oldCatalogMeta.catalogId}"
+        + " AND catalog_name = #{oldCatalogMeta.catalogName}"
+        + " AND metalake_id = #{oldCatalogMeta.metalakeId}"
+        + " AND type = #{oldCatalogMeta.type}"
+        + " AND provider = #{oldCatalogMeta.provider}"
+        + " AND catalog_comment = #{oldCatalogMeta.catalogComment}"
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/storage/relational/mapper/CatalogMetaBaseProvider.java:142]: I think you should merge back that fix you made to here.

File: core/src/main/java/org/apache/gravitino/storage/relational/mapper/CatalogMetaBaseProvider.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+ */
+
+package org.apache.gravitino.storage.relational.mapper;
+
+import static org.apache.gravitino.storage.relational.mapper.CatalogMetaMapper.TABLE_NAME;
+
+import org.apache.gravitino.storage.relational.po.CatalogPO;
+import org.apache.ibatis.annotations.Param;
+
+public class CatalogMetaBaseProvider {
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/storage/relational/mapper/CatalogMetaBaseProvider.java:27]: I would suggest rename this class to `CatalogMetaBaseSQLProvider`

File: core/src/main/java/org/apache/gravitino/storage/relational/mapper/CatalogMetaBaseProvider.java (hunk 4 above)
Comment (yuqi1129) [reply core/src/main/java/org/apache/gravitino/storage/relational/mapper/CatalogMetaBaseProvider.java:142]: OK, I will rebase the main branch.

File: core/src/main/java/org/apache/gravitino/storage/relational/mapper/CatalogMetaBaseProvider.java (hunk 3 above)
Comment (yuqi1129) [reply core/src/main/java/org/apache/gravitino/storage/relational/mapper/CatalogMetaBaseProvider.java:40]: It's a typo and won't affect the SQL string, I'll fix it.

File: core/src/main/java/org/apache/gravitino/storage/relational/mapper/CatalogMetaBaseProvider.java (hunk 5 above)
Comment (yuqi1129) [reply core/src/main/java/org/apache/gravitino/storage/relational/mapper/CatalogMetaBaseProvider.java]: done

File: core/src/main/java/org/apache/gravitino/storage/relational/mapper/CatalogMetaSQLProviderFactory.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+        SqlSessionFactoryHelper.getInstance()
+            .getSqlSessionFactory()
+            .getConfiguration()
+            .getDatabaseId();
+
+    JDBCBackendType jdbcBackendType = JDBCBackendType.fromString(databaseId);
+    return METALAKE_META_SQL_PROVIDER_MAP.get(jdbcBackendType);
+  }
+
+  static class CatalogMetaMySQLSQLProvider extends CatalogMetaBaseSQLProvider {}
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/storage/relational/mapper/CatalogMetaSQLProviderFactory.java:48]: I don't think duplicating "SQL" is a good name, just "CatalogMetaMySQLProvider" is enough, Also for `CatalogMetaH2SQLProvider`, `CatalogMetaH2Provider` is enough.

**Review by jerryshao — APPROVED** [review]

**Pull request merged**